# Ecoli60
## Synonymous mutations and neutral evolution in bacteria

This repository contains the files necessary to reproduce the results reported in our analysis of the role of synonymous in purifying selection. The *bin* folder contains the code used to generate analysis and figures. It's based on R software (version 3.5.1) and uses the libraries cowplot, doParallel, dplyr, ecoli2.db, ggplot2, grDevices, grid, gridExtra, gridGraphics, lattice, org.EcK12.eg.db, reshape2, RMySQL, scales, stats, and stringr.

The *data_files*  folder contains data generated by the pipeline. You can start running the original pipeline following the instructions found in [Time-resolved metagenomic sequencing of Lenski's long-term evolution experiment with Escherichia coli ](https://github.com/benjaminhgood/LTEE-metagenomic). Alternatively, you can start [our pipeline](#any) using pre-processed files find in *preProcess* folder.

<a name="any"></a>
###  Our pipeline  
To run our pipeline, go to [bin](https://github.com/LabBiosystemUFRN/Ecoli60/tree/master/bin/) folder and run:
> 00pipeline.sh


It will execute the final tasks of the original pipeline, generate other files needed, set the R environment up, and create the paper figures.

### Monte Carlo simulation
Scripts needed to generate the Monte Carlo simulation can be found [here](https://github.com/LabBiosystemUFRN/Ecoli60/tree/master/bin/monteCarlo/).


### Python dependencies

python 2.7, 
    bz2, 
    itertools, 
    glob, 
    math, 
    matplotlib, 
    numpy, 
    os, 
    pylab, 
    scipy, 
    sys.

### R dependencies

cowplot, doParallel, dplyr, ggplot2, grDevices, grid, gridExtra, gridGraphics, lattice, reshape2, RMySQL, scales, stats, stringr, BiocManager

### R dependencies from Bioconductor

ecoli2.db, org.EcK12.eg.db, ecoli2.db
